code/dictation.py

```diff
@@ -143,41 +171,98 @@ def __init__(self):
-    def format(self, text):
-        if needs_space_between(self.before, text):
+    def format(self, text, auto_cap=True):
+        if not self.force_no_space and needs_space_between(self.before, text):
             text = " " + text
-        text, self.state = auto_capitalize(text, self.state)
+        self.force_no_space = False
+        if auto_cap:
+            text, self.state = auto_capitalize(text, self.state)
+        if self.force_cap:
+            text = format_first_letter(text, lambda s: s.capitalize())
+            self.force_cap = False
+        if self.force_no_cap:
+            text = format_first_letter(text, lambda s: s.lower())
+            self.force_no_cap = False
         self.before = text or self.before
         return text
 
+def format_first_letter(text, formatter):
+    i = -1
+    for i, c in enumerate(text):
+        if c.isalpha():
+            break
+    if i >= 0 and i < len(text):
+        text = text[:i] + formatter(text[i]) + text[i+1:]
+    return text
+
 dictation_formatter = DictationFormat()
 ui.register("app_deactivate", lambda app: dictation_formatter.reset())
 ui.register("win_focus", lambda win: dictation_formatter.reset())
 
+def reformat_last_utterance(formatter):
+    text = actions.user.get_last_phrase()
+    actions.user.clear_last_phrase()
+    text = formatter(text)
+    actions.user.add_phrase_to_history(text)
+    actions.insert(text)
+
 @mod.action_class
 class Actions:
     def dictation_format_reset():
         """Resets the dictation formatter"""
         return dictation_formatter.reset()
 
+    def dictation_format_cap():
+        """Sets the dictation formatter to capitalize"""
+        dictation_formatter.force_cap = True
+        dictation_formatter.force_no_cap = False
+
+    def dictation_format_no_caps():
+        """Sets the dictation formatter to not capitalize"""
+        dictation_formatter.force_no_cap = True
+        dictation_formatter.force_cap = False
+
+    def dictation_format_no_space():
+        """Sets the dictation formatter to not prepend a space"""
+        # Typically helpful to reset capitalization as well.
+        dictation_formatter.reset()
+        dictation_formatter.force_no_space = True
+
+    def dictation_reformat_cap():
+        """Capitalizes the last utterance"""
+        reformat_last_utterance(lambda s: format_first_letter(s, lambda c: c.capitalize()))
+
+    def dictation_reformat_no_caps():
+        """Lowercases the last utterance"""
+        reformat_last_utterance(lambda s: format_first_letter(s, lambda c: c.lower()))
+
+    def dictation_reformat_no_space():
+        """Removes space before the last utterance"""
+        reformat_last_utterance(lambda s: s[1:] if s.startswith(" ") else s)
+
     def dictation_insert_raw(text: str):
         """Inserts text as-is, without invoking the dictation formatter."""
-        dictation_formatter.pass_through(text)
-        actions.insert(text)
+        actions.user.dictation_insert(text, auto_cap=False)
```

Comment:
This change could be dropped without affecting other pieces. I found this made formatters easier to use within dictation because you still get the auto-spacing.

Reply:
@splondike Note this is a breaking change, I can revert if you'd like. It's not used in many places, though.